[collectd 6] disk plugin: Align metrics with OpenTelemetry recommendations.

octo opened the pull request:
* Rename labels to `system.device` and `disk.io.direction`.
* Rename `system.disk.time` to `system.disk.operation_time`.
* Add descriptions and units to all metric families.
* Add the "utilization" metric to FreeBSD.

ChangeLog: Disk plugin: The metric schema has been aligned with OpenTelemetry semantic conventions.

eero-t commented on src/disk.c:
```diff
@@ -981,12 +1006,8 @@ static int disk_read(void) {
-
-      m.value.gauge = ((diff_io_time / (double)interval) * 100.0);
-      metric_family_metric_append(&fam_disk_utilization, m);
+      m.value.gauge = ((gauge_t)diff_io_time) / ((gauge_t)interval);
```
div-by-zero if interval is zero?

octo replied:
I think we can assume that the interval is non-zero. Otherwise I think a lot will break in the daemon. We divide by the interval somewhat frequently to calculate rates.

eero-t commented on src/disk.c:
```diff
@@ -1217,20 +1238,20 @@ static int disk_read(void) {
 
-    m.value.counter = drives[i].time_sec * 1000 + drives[i].time_usec / 1000;
-    metric_family_metric_append(&fam_disk_io_time, m);
+    m.value.counter = drives[i].time_sec + drives[i].time_usec / 1000000;
+    metric_family_metric_append(&fam_ops_time, m);
```
This metric family name change does not correspond with changes at start of the `disk_read()`.

octo replied:
Not sure I understand you correctly. Did you mean that this is in seconds while other systems report in us? I have changed all implementations to report us.

eero-t replied:
I mean that this change is:
* `fam_disk_io_time` -> `fam_ops_time`

Whereas other changes in the file were:
* `fam_disk_io_time` ->  `fam_disk_io_time`  (no fam name change)
* `fam_time` -> `fam_ops_time`

eero-t reviewed (commented):
I think it would be better to handle https://github.com/collectd/collectd/pull/4242 in `main`, and merge all `main` changes to `disk.c` before this PR.

Reporting (some) time units as seconds (`s`) is IMHO not good while collectd counters are integers i.e. do not support subsecond timings.  Meaning that one does not see from counters very small usage spikes i.e. is disk completely idle or not.

Also, in general time manipulation for values appended to `fam_ops_time` is confusing as they can be multiplied or divided by different values, and variable names not having any indication of indication of their time units.

octo commented on src/disk.c:
```diff
@@ -382,54 +388,70 @@ static signed long long dict_get_value(CFDictionaryRef dict, const char *key) {
       .name = "system.disk.operations",
+      .help = "Read and write operations performed by the disk. If multiple "
+              "operations are merged into one by the disk, they will be "
+              "accounted for separately.",
```
I like your wording. I have changed "disk" to "I/O subsystem" because I'm not sure if the kernel, the disk, or both merge operations.

octo commented on src/disk.c:
```diff
@@ -382,54 +388,70 @@ static signed long long dict_get_value(CFDictionaryRef dict, const char *key) {
+              "or I/O merge by the number of I/Os in progress times the number "
+              "of milliseconds spent doing I/O since the last update of this "
+              "field.  This can provide an easy measure of both I/O completion "
+              "time and the backlog that may be accumulating.",
```
Yeah, this quite a never-ending sentence. I copied it from the Kernel docs. I added some commas to make it easier to digest.

octo commented:
Hey @eero-t, thanks for the review!

>     * I think it would be better to handle [disk: add /proc/diskstats fields 15 to 20 in KERNEL_LINUX #4242](https://github.com/collectd/collectd/pull/4242) in `main`, and merge related `main` changes to `disk.c` before this PR.

The v6 code is so different that I don't think resolving the resulting merge conflict is any easier than reimplementing the feature. Writing the metric families is more effort than plumbing the counters through.

>     * Reporting (some) time units as seconds (`s`) is IMHO not good while collectd counters are integers i.e. do not support subsecond timings.  Meaning that one does not see from counters very small usage spikes i.e. is disk completely idle or not.

That was a mistake, thanks for your diligence! I have changed all platforms to report microseconds, and changed the metric family unit to reflect that.

>     * Time manipulation for values appended to `fam_ops_time` is confusing as they can be multiplied or divided by different values, and variable names do not indicate their time units (like they IMHO should).

I have added appropriate suffixes to all variable and field names.

eero-t commented on src/disk.c:
```diff
@@ -1059,26 +1080,18 @@ static int disk_read(void) {
+    metric_family_append(&fam_ops, direction_label, write_direction,
                          (value_t){.counter = kio.KIO_WOPS}, &m);
 
-    if (strncmp(ksp[i]->ks_class, "disk", 4) == 0) {
-      /* FIXME: Convert this to microseconds if necessary */
-      metric_family_append(&fam_time, "direction", "read",
-                           (value_t){.counter = kio.KIO_RTIME}, &m);
-      metric_family_append(&fam_time, "direction", "write",
-                           (value_t){.counter = kio.KIO_WTIME}, &m);
-    }
-
```
You commented in commit removing these Solaris metrics, them being run & wait, instead of read & write => maybe they should now use `fam_disk_io_time`?

Note: wait time is important disk metric.  It tells how much things are stalling because there's too much to write / disk is too slow.

octo replied:
I have added a few metrics for Solaris based on the kstat docs.

eero-t reviewed (changes requested):
Value wrap-up calculations in `disk.c` are broken, see: `grep UINT_MAX disk.c`

They deduct 64-bit (signed) `derive_t` value from 32-bit (unsigned) `UINT_MAX` value.

Maybe collectd should have `DERIVE_T_MAX` define for this (both in v5 & v6 branches)?

eero-t reviewed (approved):
Approved.  I'll file separate bug about  the UINT_MAX vs derive_t issue, as it's already in the v5 version of disk plugin, and not introduced by these changes.

octo commented:
> Value wrap-up calculations in `disk.c` are broken, see: `grep UINT_MAX disk.c`
> 
> They deduct 64-bit (signed) `derive_t` value from 32-bit (unsigned) `UINT_MAX` value.
> 
> Maybe collectd should have `DERIVE_T_MAX` define for this (both in v5 & v6 branches)?

That's an excellent callout, thank you! We actually have a function that is doing this right, `counter_diff()`. I ended up refactoring a lot of the Linux code; being one of the oldest pieces of code in collectd it was quite in need of a facelift.

After a lot of fiddling I figured out that we calculated the average time per operation … just for that variable never to be read. I think the existing metrics are much better at conveying useful information about the disk, so I remove the calculcation and made the plugin a good deal simpler.

We also unnecessarily stored a bunch of stuff in global state, which I have now removed.

eero-t commented:
Gah, read Linux diskstats documentation, and counters use _native_ unsigned long (see the new bug).

I.e. they are indeed 64-bit, but only on 64-bit host.  For 32-bit hosts, UINT_MAX was actually the correct maximum value.

eero-t commented:
Nice, `counter_diff()` already handles transparently unsigned 32-bit vs 64-bit wrap up => it will fix #4245 for v6.

The very best changes are ones that remove lots of code, and still manage retain the (relevant) functionality. :-)

On (very) quick check of the changes, they seemed OK, i.e. from my side this still seems fine for merging.

eero-t commented:
As to adding tests for `disk.c` (in another PR)...

I think that could be done for the Linux functionality by extracting that from `disk_read()` to another function, one that takes path to `/proc/diskstats` as an argument.   Test code can then ask it to read pre-prepared regular file(s) instead, to verify that reported metric updates are as expected, and to verify that invalid file content is handled gracefully.

(CPU plugin test coverage could be similarly extended to parsing of `/proc/stats` file.  At that point some helper functions for dealing with temporary test files could be useful.)

octo commented:
> As to adding tests for `disk.c` (in another PR)...

I think we need to separate I/O and parsing. Maybe a schema along the lines of:

```c
int parse(char const *data, metric_family_t families[]);

int read_callback(void) {
  char *buffer = read_text_file_contents();
  parse(buffer);
}
```

This will allow us to test `parse()` without using I/O.

We should also split plugins into multiple files, one for each operating system we support. E.g. `disk.c`, `disk_linux.c`, `disk_freebsd.c` etc. The current approach is hard to read and hard to maintain.
